modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c

```diff
@@ -0,0 +1,576 @@
+    ngx_int_t                      total;
+    ngx_uint_t                     i, p;
+    ngx_http_upstream_rr_peer_t  *peer, *best;
+
+    best = NULL;
+    p = 0;
+    total = 0;
+    for (peer = peers->peer, i = 0; peer; peer = peer->next, i++) {
+        peer->current_weight += peer->effective_weight;
+        total += peer->effective_weight;
+
+        if (best == NULL || peer->current_weight > best->current_weight) {
+            best = peer;
+            p = i;
+        }
+    }
+
+    *rpeer = best;
+    if (best == NULL) {
+        return NGX_ERROR;
+    }
+
+    best->current_weight -= total;
+
+    return p;
+}
+
+
+static ngx_http_upstream_rr_peer_t *
+ngx_http_upstream_get_vnswrr(ngx_http_upstream_vnswrr_peer_data_t  *vnsp)
+{
+    time_t                                  now;
+    uintptr_t                               m;
+    ngx_int_t                               rindex;
+    ngx_uint_t                              i, n, p, flag, begin_number;
+    ngx_http_upstream_rr_peer_t            *peer, *best;
+    ngx_http_upstream_rr_peers_t           *peers;
+    ngx_http_upstream_rr_vpeers_t          *vpeers;
+    ngx_http_upstream_rr_peer_data_t       *rrp;
+    ngx_http_upstream_vnswrr_srv_conf_t    *uvnscf;
+
+    now = ngx_time();
+
+    best = NULL;
+
+#if (NGX_SUPPRESS_WARN)
+    p = 0;
+#endif
+
+    rrp = &vnsp->rrp;
+    peers = rrp->peers;
+    uvnscf = vnsp->uvnscf;
+    vpeers = uvnscf->vpeers;
+
+    if (uvnscf->last_number == NGX_CONF_UNSET_UINT) {
+        uvnscf->init_number = ngx_random() % peers->number;
+
+        if (peers->weighted) {
+            peer = vpeers[uvnscf->init_number].vpeer;
+
+        } else {
+            for (peer = peers->peer, i = 0; i < uvnscf->init_number; i++) {
+                peer = peer->next;
+            }
+        }
+
+        uvnscf->last_number = uvnscf->init_number;
+        uvnscf->last_peer = peer;
+    }
+
+    if (peers->weighted) {
+        /* batch initialization vpeers at runtime. */
+        if (uvnscf->vnumber != peers->total_weight
+            && (uvnscf->last_number + 1 == uvnscf->vnumber))
+        {
+            n = peers->total_weight - uvnscf->vnumber;
+            if (n > peers->number) {
+                n = peers->number;
+            }
+
+            for (i = uvnscf->vnumber; i < n + uvnscf->vnumber; i++) {
+                rindex = ngx_http_upstream_get_rr_peer(peers, &peer);
+                if (rindex == NGX_ERROR) {
+                    ngx_log_error(NGX_LOG_ERR, ngx_cycle->log, 0,
+                                  "get rr peer is null in upstream \"%V\" ",
+                                  peers->name);
+                    if (i != 0) {
+                        i--;
+                    }
+
+                    continue;
+                }
+
+                vpeers[i].vpeer = peer;
+                vpeers[i].rindex = rindex;
+            }
+
+            uvnscf->vnumber = i;
+        }
+
+        begin_number = (uvnscf->last_number + 1) % uvnscf->vnumber;
```

Comment:
Maybe we don't need to `% uvnscf->vnumber` here? There is already a `% uvnscf->vnumber` in the for loop below.

Reply:
This is necessary.  Because the initial value of `begin_number` is `ngx_random() % peers->number ` , But `uvnscf->vnumber <= peers->number`.